File: js/task3/S3_PIPE_03.js
// S3_PIPE_03 — String normalization pipeline

function pipe(...fns) {
  return (input) => fns.reduce((v, fn) => fn(v), input);
}

const trim = (s) => s.trim();
const lower = (s) => s.toLowerCase();
const collapseSpaces = (s) => s.replace(/\s+/g, " ");

const normalize = pipe(trim, lower, collapseSpaces);

// Tests
console.log(normalize("   HeLLo   WoRLD   ")); // "hello world"
console.log(normalize("A   B    C"));          // "a b c"
console.log(normalize("   ONE\tTWO  THREE ")); // "one two three"

File: python/task3c/S3_PIPE_03.py
# S3_PIPE_03 — String normalization pipeline

import re

def pipe(*fns):
    def runner(x):
        for fn in fns:
            x = fn(x)
        return x
    return runner


def main():
    normalize = pipe(
        lambda s: s.strip(),
        lambda s: s.lower(),
        lambda s: re.sub(r"\s+", " ", s),
    )

    text = " Ala Ma Kota "
    print("original:", repr(text))
    print("normalized:", repr(normalize(text)))


if __name__ == "__main__":
    main()
